examples/crd_derive_multi.rs

```diff
@@ -48,42 +210,56 @@ async fn main() -> anyhow::Result<()> {
+    // create a v1 resource
     let v1m = v1::ManyDerive::new("old", v1::ManyDeriveSpec {
         name: "i am old".into(),
         oldprop: 5,
     });
     let oldvarv1 = v1api.patch("old", &ssapply, &Patch::Apply(&v1m)).await?;
     info!("old instance on v1: {:?}", oldvarv1.spec);
     let oldvarv2 = v2api.get("old").await?;
-    info!("old instance on v2 truncates: {:?}", oldvarv2.spec);
+    info!("old instance on v2 not truncated: {:?}", oldvarv2.spec);
+    assert_eq!(oldvarv2.spec.newprop, 5); // no data loss
 
     // create a v2 version
     let v2m = v2::ManyDerive::new("new", v2::ManyDeriveSpec {
         name: "i am new".into(),
-        extra: Some("hi".into()),
+        newprop: 4,
     });
     let newvarv2 = v2api.patch("new", &ssapply, &Patch::Apply(&v2m)).await?;
-    info!("new instance on v2 is force downgraded: {:?}", newvarv2.spec); // no extra field
-    let cannot_fetch_as_old = v1api.get("new").await.unwrap_err();
-    info!("cannot fetch new on v1: {:?}", cannot_fetch_as_old);
-
-    // apply schema upgrade
-    apply_crd(client.clone(), merge_crds(all_crds, "v2")?).await?;
+    info!(
+        "new instance on v2 was force downgraded on storage: {:?}",
+        newvarv2.spec
+    );
+    assert_eq!(newvarv2.spec.newprop, 4); // no data loss
+    let can_fetch_as_old = v1api.get("new").await.unwrap();
+    info!(
+        "fetched new instance using old api version: {:?}",
+        can_fetch_as_old
+    );
 
-    // nothing changed with existing objects without conversion
-    //let oldvarv1_upg = v1api.get("old").await?;
-    //info!("old instance unchanged on v1: {:?}", oldvarv1_upg.spec);
-    //let oldvarv2_upg = v2api.get("old").await?;
-    //info!("old instance unchanged on v2: {:?}", oldvarv2_upg.spec);
+    // as you can see, proper conversion allows us to access the same object under different
+    // apiVersions without any problems. For example, you may have custom controller which uses v1 api
+    // and admission webhook which uses v2.
 
-    // re-apply new now that v2 is stored gives us the extra properties
-    let newvarv2_2 = v2api.patch("new", &ssapply, &Patch::Apply(&v2m)).await?;
-    info!("new on v2 correct on reapply to v2: {:?}", newvarv2_2.spec);
+    // Now imagine that we upgraded all our clients to the v2 API
+    // and we want to drop support for the legacy v1 API (after all you see how many code we had to write
+    // for conversion).
 
+    // at first, we need to promote `v2` version as stored (so that all new or updated resources are persisted
+    // in the cluster storage using new version).
+    info!("Selecting v2 as storage version");
+    apply_crd(client.clone(), merge_crds(all_crds.clone(), "v2")?).await?;
+    // here we use `migrate_resources` utility function to migrate all previously stored objects.
+    info!("Running storage migration");
+    let crds: Api<CustomResourceDefinition> = Api::all(client.clone());
+    crds.migrate_resources(&crd2.name_unchecked());
```

Comment:
Q: is it possible for crd2 and crd1 to have different names? I don't think so, because then it would be a different CRD. If not, we don't even need to pass in a parameter here. We can just call `name_unchecked` internally.

Reply:
> I don't think so, because then it would be a different CRD

I don't think so too. AFAIK, CRDs must be named as "${KIND}.${GROUP}".

> If not, we don't even need to pass in a parameter here

What do you mean? Change from `crd_name: &str` to `crd: &CustomResourceDefinition`?

Reply:
Ah, I thought there was something smart here, but I confused the CRs and the CRDs in the example code.